cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ServiceHttpServer.java

```diff
@@ -294,16 +296,27 @@ public void run() {
   }
 
   private void initHandler(final HttpServiceHandler handler, final BasicHttpServiceContext serviceContext) {
+    TransactionControl txCtrl = Transactions.getTransactionControl(TransactionControl.IMPLICIT, Object.class,
+                                                                   handler, "initialize", HttpServiceContext.class);
     ClassLoader classLoader = setContextCombinedClassLoader(handler);
-    DataFabricFacade dataFabricFacade = dataFabricFacadeFactory.create(program,
-                                                                       serviceContext.getDatasetCache());
     try {
-      dataFabricFacade.createTransactionExecutor().execute(new TransactionExecutor.Subroutine() {
-        @Override
-        public void apply() throws Exception {
-          handler.initialize(serviceContext);
-        }
-      });
+      if (TransactionControl.IMPLICIT == txCtrl) {
+        // it is not obvious but we can't use the serviceContext.execute() here. This is because multiple instances
+        // of the handler may initialize at the same time, and they may conflict with each other. The serviceContext
+        // initialize() does not have a retry strategy, whereas the dataFabricFacade creates an executor with retry.
+        // TODO: make this cleaner and do it for other program types, too.
+        dataFabricFacadeFactory
+          .create(program, serviceContext.getDatasetCache())
+          .createTransactionExecutor()
+          .execute(new TransactionExecutor.Subroutine() {
```

Comment:
Can we use the `Transactions.createTransactionalWithRetry` instead? How is it different than using the `TransactionExecution.execute`?